### PR TITLE
Removing support for asynchronous server requests

### PR DESCRIPTION
**Pull request by nikhiljindal**
As per discussion in https://github.com/GoogleCloudPlatform/kubernetes/pull/3697.
This is to stop exposing Operations to users, so that we can remove /operations.

### CODE REVIEW

**Comment by nikhiljindal** [comment]
Trying to fix broken tests

**Comment by nikhiljindal** [comment]
Travis is green!

**Comment by smarterclayton** [comment]
Do you have a clean e2e run?  Just in case

**Comment by erictune** [comment]
IIUC, all operations are now synchronous?  Would it make sense to remove the sync query parameter from everywhere it is being used, perhaps in a follow on PR?  And make any documentation reflect that it is accepted but ignored?  And is it now the case that 202 won't be returned?  Should there be a cleanup to remove mentions of 202 from the code?  And do you need to run hack/build-ui.sh so that the ui will show any changes in the REST interface?

**Comment by bgrant0607** [comment]
I agree with @erictune. We might as well remove uses of `sync` in this PR.
```
./pkg/client/request_test.go:   fakeHandler.ValidateRequest(t, "/api/v1beta1/foo/bar/baz?labels=name%3Dfoo&sync=true&timeout=1s", "POST", &tmpStr)
./pkg/client/request.go:// parameter. Ignored if sync=false.
./pkg/apiserver/apiserver_test.go:      request, err := http.NewRequest("POST", server.URL+"/prefix/version/ns/other/foo?sync=true", bytes.NewBuffer(data))
./pkg/apiserver/apiserver_test.go:      itemOut := expectApiStatus(t, "POST", server.URL+"/prefix/version/foo?sync=true&timeout=4ms", data, http.StatusAccepted)
./pkg/apiserver/resthandler.go://    sync=[false|true] Synchronous request (only applies to create, update, delete operations)
./pkg/apiserver/resthandler.go://    timeout=<duration> Timeout for synchronous requests, only applies if sync=true
./test/integration/auth_test.go:var syncFlags = "?sync=true&timeout=60s"
```
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/client/request.go#L42
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/client/request.go#L110
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/client/request.go#L364

**Comment by erictune** [comment]
it is used in

test/integration/auth_test.go

**Comment by smarterclayton** [comment]
+1

**Comment by nikhiljindal** [comment]
Thanks Eric and Brian for the code pointers.
Have updated the code to delete all references of sync and 202 status code.

Also ran hack/test-go and hack/e2e-test locally to ensure that all is fine.

**Comment by smarterclayton** [comment]
We can

File: pkg/apiserver/resthandler.go [hunk 1]
```diff
@@ -308,13 +305,9 @@ func (h *RESTHandler) handleRESTStorage(parts []string, req *http.Request, w htt
...
 // createOperation creates an operation to process a channel response.
-func (h *RESTHandler) createOperation(out <-chan RESTResult, sync bool, timeout time.Duration, onReceive func(RESTResult)) *Operation {
+func (h *RESTHandler) createOperation(out <-chan RESTResult, timeout time.Duration, onReceive func(RESTResult)) *Operation {
 	op := h.ops.NewOperation(out, onReceive)
-	if sync {
-		op.WaitFor(timeout)
-	} else if h.asyncOpWait != 0 {
-		op.WaitFor(h.asyncOpWait)
-	}
+	op.WaitFor(timeout)
```
Comment (smarterclayton) [inline pkg/apiserver/resthandler.go]: Wouldn't it just be easier to remove operation completely, and replace it with the code from op.WaitFor with a simple timeout?  We already have the http level max client timeout on the server, and I'd like to remove the extra locking operations on Operations and the two go routines it creates.

Comment (nikhiljindal) [inline pkg/apiserver/resthandler.go]: I am trying to break this into multiple small manageable changes.
This PR is to make all operations synchronous and hence never return operations to users.
Then, in next PR, we can delete the /operations endpoint.
Once these 2 are submitted, we can delete operations all together in a final PR.

Does that sound good or do you think its better to do it all at once?

Comment (smarterclayton) [inline pkg/apiserver/resthandler.go]: That's fine - I just wasn't sure what the plan was.

----- Original Message -----

**Comment by smarterclayton** [comment]
LGTM

File: pkg/api/types.go [hunk 2]
```diff
@@ -926,6 +914,12 @@ const (
 	// Status code 422
 	StatusReasonInvalid StatusReason = "Invalid"
 
+	// StatusReasonTimeout means that the request could not be completed within the given time.
```
Comment (bgrant0607) [inline pkg/api/types.go]: IIUC, the lists need to match in the versioned types.go files. I don't understand why this list contains so many constants that the others do not, and why the others contain at least one that this does not.

@smarterclayton Is this just a bug?

Comment (smarterclayton) [inline pkg/api/types.go]: Yes, I don't think we have removed / changed any constants across versions beyond the lowerCase to CamelCase change, so they should all exist consistently.

Comment (nikhiljindal) [inline pkg/api/types.go]: Yes. It wasnt clear to me why some of the constants are duplicated in versioned types.
It didnt look like they are needed there and hence I didnt add them.

Comment (smarterclayton) [inline pkg/api/types.go]: Constants should always be duplicated in the event we have to covert them - someone who depends on constant from specific version should be able to rely on that in code.

File: pkg/apiserver/resthandler.go [hunk 3]
```diff
@@ -144,12 +143,10 @@ func curry(f func(runtime.Object, *http.Request) error, req *http.Request) func(
...
 // Returns 404 if the method/pattern doesn't match one of these entries
 // The s accepts several query parameters:
-//    sync=[false|true] Synchronous request (only applies to create, update, delete operations)
-//    timeout=<duration> Timeout for synchronous requests, only applies if sync=true
+//    timeout=<duration> Timeout for synchronous requests
 //    labels=<label-selector> Used for filtering list operations
 func (h *RESTHandler) handleRESTStorage(parts []string, req *http.Request, w http.ResponseWriter, storage RESTStorage, namespace, kind string) {
 	ctx := api.WithNamespace(api.NewContext(), namespace)
-	sync := req.URL.Query().Get("sync") == "true"
 	timeout := parseTimeout(req.URL.Query().Get("timeout"))
```
Comment (bgrant0607) [inline pkg/apiserver/resthandler.go]: Please add a TODO to document the timeout parameter.

**Comment by bgrant0607** [comment]
Did you run e2e?

File: pkg/apiserver/resthandler.go (hunk 3 above)
Comment (nikhiljindal) [inline pkg/apiserver/resthandler.go]: Done.

**Comment by nikhiljindal** [comment]
Yes. I had run both hack/test-go and hack/e2e-test locally to ensure that all is fine.

**Comment by smarterclayton** [comment]
If no further comments I'll merge in the morning

File: pkg/api/types.go (hunk 2 above)
Comment (bgrant0607) [inline pkg/api/types.go]: I'm fine if this is addressed in a separate PR.

@nikhiljindal Please file an issue if you don't fix it in this PR.

**Comment by bgrant0607** [comment]
LGTM, too

**Pull request merged**